### ISSUE TITLE
alpha 3 hangs at installing base os

### ISSUE THREAD

**Issue by Ragebone** [justinclift/freenas-infiniband#4]
On my try to install the alpha 3 it hangs after it finishes the first of 5 base os install steps.

Console output is:
`Installing base-os (1 of 5)`
`.....10...20 ...  .. 100 `

nothing happens after that.

**Comment by justinclift** [comment]
Interesting.  I had the same thing seem to happen here on one system, though not on a different one.

On the system it seemed to hang on, I left it overnight and upon returning it had finished.

So, I suspect there's something in the installer that's interacting badly with your system, through no fault of your own.

Any chance you could leave it running for a few hours and see if it completes?

**Comment by Ragebone** [comment]
Thank you for answering so fast.
This was just a test install for me, i have QLogic QLE7340 and i won't run it all night just to figure out if my cards are working somehow.
Maybe that's the reason? Unsupported IB Cards?
Hardware is a Supermicro X9dri-f with one e5 2670 and currently 4Gb of ram.

It looks like, the alpha2_IB_eth has the same problem.
I'll try 8Gb now.

**Comment by justinclift** [comment]
Yeah, I have no idea what the cause was for my setup either.  I installed FreeBSD on the box that had the same problem without issues. ^\o/^

People mentioned having the same problem in the FreeNAS developers forum around the same time too, for their own nightly builds (eg non-IB related)  So I'm guessing it's just our bad luck in capturing a code base which had a bug at the time.

With the QLogic QLE7340 card, there's pretty much zero chance it'll work.  FreeNAS is based upon FreeBSD and uses FreeBSD drivers.  As far as I know there's no FreeBSD driver for those cards. ☹️ 

If you do find one somewhere, point me at it as I'm happy to make an "alpha 4" build including it. :smile:

**Comment by Ragebone** [comment]
Those cards seem to work off the base ib_umad / ib_mad librarys / modules.
OFED if the referenced software by QLogic nad it uses basically teh same librarys and modules, so i think that it will be able to work with those, and because OFED is implemented in the freeBSD kernel since V9.3 i have the hope of it working.

On my Linux Mint install i had those cards be detected and all the stats and configs did work. 
I'm still a noob and i didn't get them connected .. but i'll get that somehow.

**Comment by justinclift** [comment]
Good news about the install finishing. 😀 

Ahhh yeah, I'd expect them to work in Linux.  On linux, you probably just need to run the "OpenSM" daemon so the cards get assigned an IB address and can see each other.

With FreeNAS... I have no idea how that'll play out.  Um, good luck and let us know if you get it working? 😉

**Comment by Ragebone** [comment]
Thanks for the suggestion.
Before i try that, i still need to get the counterpart working. Your alpha3 at the moment.

ibstat crashes with : 
`ibpanic:` [6561] main: stat of IB device 'mthca0' failed: (m)`

**Comment by justinclift** [comment]
Damn.  I have no idea why it's trying the `mthca` driver, that's specifically for Mellanox Infinihost III cards (ancient).

**Comment by Ragebone** [comment]
Well, do you allow recapping what i have researched so far ?

I have listed all the loaded kernel modules and there are no familiar IB modules visible to me.
output of kldstat just to prove that the kernel modules aren't there:
`
 1   83 0xffffffff80200000 1957000  kernel                                                                                          
 2    1 0xffffffff81e44000 ffd8c    ispfw.ko                                                                                        
 3    1 0xffffffff821e6000 106b2    geom_mirror.ko                                                                                  
 4    1 0xffffffff821f7000 4691     geom_stripe.ko                                                                                  
 5    1 0xffffffff821fc000 ffe9     geom_raid3.ko                                                                                   
 6    1 0xffffffff8220c000 ec6a     geom_raid5.ko                                                                                   
 7    1 0xffffffff8221b000 5781     geom_gate.ko                                                                                    
 8    1 0xffffffff82221000 4a3a     geom_multipath.ko                                                                               
 9    1 0xffffffff82226000 5734     fdescfs.ko                                                                                      
10    1 0xffffffff8222c000 89c      dtraceall.ko                                                                                    
11   10 0xffffffff8222d000 3ad56    dtrace.ko                                                                                       
12    1 0xffffffff82268000 466e     dtmalloc.ko                                                                                     
13    1 0xffffffff8226d000 225c     dtnfscl.ko                                                                                      
14    1 0xffffffff82270000 6403     fbt.ko                                                                                          
15    1 0xffffffff82277000 579c2    fasttrap.ko                                                                                     
16    1 0xffffffff822cf000 4a00     lockstat.ko                                                                                     
17    1 0xffffffff822d4000 1630     sdt.ko                                                                                          
18    1 0xffffffff822d6000 d8f0     systrace.ko                                                                                     
19    1 0xffffffff822e4000 d4c5     systrace_freebsd32.ko                                                                           
20    1 0xffffffff822f2000 4dd0     profile.ko                                                                                      
21    1 0xffffffff822f7000 7fbf     ipmi.ko                                                                                         
22    1 0xffffffff822ff000 ed9      smbus.ko                                                                                        
23    1 0xffffffff82300000 1a66e    hwpmc.ko                                                                                        
24    1 0xffffffff8231b000 c2d4     t3_tom.ko                                                                                       
25    2 0xffffffff82328000 42ed     toecore.ko                                                                                      
26    1 0xffffffff8232d000 12b9a    t4_tom.ko                                                                                       
27    1 0xffffffff82340000 3620     ums.ko                                                                                          
28    2 0xffffffff82344000 2b82     vboxnetflt.ko                                                                                   
29    2 0xffffffff82347000 4e416    vboxdrv.ko                                                                                      
30    1 0xffffffff82396000 41f0     ng_ether.ko                                                                                     
31    1 0xffffffff8239b000 3fd4     vboxnetadp.ko   
`
this states OFED beeing in freebsd since 2011. 9.0 and later.
https://wiki.freebsd.org/InfiniBand
Have you compiled the Freenas Alpha with OFED enabled like in this wiki post?
`Add WITH_OFED='yes' to /etc/src.conf.`

This states somewhere that this is for those cards.  ctrl+f  7340 
http://filedownloads.qlogic.com/files/ms/75852/Users_Guide_QLogic_OFED_Host_Software_15_A.pdf

I don't want to annoy you and its getting late here in Germany, so never the less, thank you very much for your help!

**Comment by justinclift** [comment]
Looking at the kldstat output (it's came through via email), I'm not seeing any hardware driver that looks QLogic related. ☹️ 

With OFED, it was definitely enabled both in the kernel (for the kernel bits), and in `/etc/src.conf` so the IB system utils were built.

As a thought I'm not sure if the parts of OFED that were imported into FreeBSD included the hardware drivers for QLogic.  I ***suspect*** (but don't know) they wouldn't have been, as the work was done by Mellanox (competitor to QLogic).

If you have the time, it's probably a decent idea to ask about these cards on the FreeBSD Infiniband mailing list:

&nbsp; &nbsp; https://lists.freebsd.org/mailman/listinfo/freebsd-infiniband

**Comment by Ragebone** [comment]
you are right with that there are no specific module or driver installed for the qlogic cards.
But i think that those cards don't need that!

When i tried to get the cards working on linux Mint, i'll use synaptics packagemanager, and install
`infinband-diags`  // i just searched for infiniband and hoped for something.
That depends on 
`libopensm5`
`libibumad3`
`libibmad5`
// i think libibumad3  and libibmad5 add the kernel module ib_umad.

after that, i can `sudo modprobe ib_umad` to load the kernel module.
and from that point on, ibstat prints the typical card info.

That makes me assume, that the card is getting recognized and is at least partially functional.
I further assume that everything  necessary to load, recognize and hopefully use the card is contained in the ib_umad kernel module.
Output of `ibstat`
`CA 'qib0'`
	`	CA type: InfiniPath_QLE7340`
	`	Number of ports: 1`
	`	Firmware version: `
	`	Hardware version: 2`
	`	Node GUID: 0x0011750000782a46`
	`	System image GUID: 0x0011750000782a46`
	`	Port 1: ....`

Installing and using `opensm` it apears that opensm is capable of working with the card.
Launching it prints
`OpenSM 3.3.15`
`Entering DISCOVERING state`
`Using default GUID 0x11750000782a46`
`SM port is down`

At this point, i don't jet know how to proceed and all the tutorials using OFED have thrown me back a bit because i then tried to understand OFED and how that works.

Thank you for the  time and suggestions!

**Comment by justinclift** [comment]
Ahhhh.  When I look at your output there, the `CA 'qib0'` line in `ibstat` says to me it's using a driver called "qib".  It's calling it "qib0", because it's the first (and only) interface on the card.

Doing a quick online search for "qib" shows there is indeed a Linux kernel module called "qib", and yep, it's the hardware module for QLogic IB cards:

&nbsp; &nbsp; http://lxr.free-electrons.com/source/drivers/infiniband/hw/qib/

**Comment by justinclift** [comment]
For "how to proceed"... um, it depends on your goal. :smile:

If you're wanting a general file server, and definitely want/need to use those QLogic cards, you're way better off building a Linux based one as the card is supported there.  Unless you're up for writing driver software. (?)

Btw, that OpenSM bit working is good news.  If you connect that card to another box with an IB card (eg via IB cable), then the "SM port is down" message should change to say it's up and running.  With that done, enable IPoIB and you'll be able to assign IP addresses to either and and send traffic between them.

And yeah, I need sleep too.  But it's something for you to experiment with more when you have sometime. :smile:

**Comment by justinclift** [comment]
Looking through that QLogic adapter user guide PDF you linked to above, it really does seem like there's a specific "qib" driver for the adapter.  Called `ib_qib` in various places/diagrams there.  Figure 3-1 seems to illustrate things the way I understand them.

eg lowest layer is the IB adapter itself, hardware driver (ib_qib) is just above that, talking to the card, with the uMAD and other utilities one layer higher, talking to the driver.

It really, really, really :wink: seems like you'd need a FreeBSD version of the qib driver for the FreeBSD OFED bit to talk to.

**Comment by Ragebone** [comment]
Wow! 
Looking through your messages, i started googling for ib_qib and it turns out that there is a github repo.
https://github.com/01org/qib
Looks like its the right one.
And it was added to the linux kernel:
http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=f931551bafe1f10ded7f5282e2aa162c267a2e5d

And it appears. there are ways of getting a linux driver to work in FreeBSD.

So, i plan on looking further into this! 
Plan is to try raw freeBSD this evening, just to have at least tried it.
But im just a CS Student who still didn't do the c lessons, so keep the expectations low.

But back to my hardware, booted up mint on my server, and had both configured exactly the same.
Result is that opensm isn't working correctly. 
And even though both cards are connected with a 10m direct attach copper cable, with mellanox trancievers, no link lights and no change of state.
The trancievers are keyed and only fit the way they are in and it locks in place.

But you are not supposed to fix or diagnose my hardware issues, so google here i come!

**Comment by justinclift** [comment]
No worries.  As a thought, definitely check if there's newer firmware available for your card.

With Mellanox cards, it's very common to get them off eBay (etc) with old firmware.  The older firmware causes all kinds of weird issues, easily fixed by updating it.

If Qlogic cards work the same way, you're best off taking care of that right at the start. :smile:

**Comment by Ragebone** [comment]
Not directly from Qlogic or intel, but a product page.
> The QLE7340’s advanced design does not need onboard firmware or external memory, which ..

Well, i can't find any openly available download for any firmware nor instrucktions on how to flash that firmware.
EDIT: i found one mention of a Frimware update problem workaround i OFED. Could be for their switches though.


Intel has this  "true Fabric" thing, but its hidden behind logins and product activations.

**Comment by Ragebone** [comment]
I searched for "qib freebsd" and stumbled across this commit:
https://lists.freebsd.org/pipermail/svn-src-projects/2011-January/003310.html
A hughe part of the qib files went in there. It also contains the ..7322.c that contains code for the QLE7240.

This also points at a qib driver in freebsd
http://fxr.watson.org/fxr/source/ofed/drivers/infiniband/Kconfig

I'll try raw freeBSD 11 this evening.

**Comment by justinclift** [comment]
Hey wow, you might have found a way forward. :smile:

**Comment by Ragebone** [comment]
Thanks! 

two things.

I probably have a bad cable, because i got both cards working as Ethernet network devices with an IP. They could ping themselves but not the other one through the cable. No blinky lights either, and the port stays "down".
I assume Mellanox Transciever don't like Qlogic cards. Or i'm just unlucky with the cable.
 :sob:  

Back to the Software. I tried to find the driver files that your alpha 3 uses on my cards, but didn't see them anywhere. I also didn't spot any loaded or available kernel modules who would add core IB support.
something like   `ibcore` or `ib_umad, ib_mad, ibverbs,` or something like that.
Locations i searched:` /boot/kernel  /boot/driver /usr/..` i forgot that one.

Could it be that those drivers and modules get compiled into the monolithic kernel ?
If that's the case, how can i figure out if the `qib` driver is in the kernel and maybe not used for some reason? 
With the knowledge that there were parts of `qib` in FreeBSD, where have they gone?
probably not camping :smile:

Shall we open up another "issue" to direct our conversion from "off topic" to something more fitting?
i'm ok with both, for me this has been a blast! 
Thanks for that!

**Comment by justinclift** [comment]
Cool, sounds like you're actually on track to get this working.  :smile:

With the alpha3 kernel, it was built using this monolithic kernel file:

&nbsp; &nbsp; https://github.com/justinclift/freenas-infiniband/blob/FreeNAS-IB-9.10-alpha3/build/profiles/freenas9/kernel/FREENAS.amd64

And yep, no `qib` driver included.  I didn't know it existed. :wink:

If you're up for building an "alpha4" yourself (eg with "qib" driver included), you probably won't find it hard.  The build process to create FreeNAS .iso's is fairly simple:

&nbsp; &nbsp; https://github.com/freenas/freenas-build/wiki/FreeNAS-9.10---10-—-Setting-up-a-FreeNAS-build-environment

After you clone the initial freenas-build repo, add this patch to it:

&nbsp; &nbsp; https://github.com/justinclift/freenas-infiniband/commit/9117aaa127230f2f6cdb379c8eb1c99355ad29dc

That adds the IB bits I know of (OFED, Mellanox drivers, etc).  Tack an entry for the `qib` driver on too, then do the build.

That should get it done. :smile:

With the Issue, lets just keep it here.  We can change the title to something like "Add support for QLogic Infiniband cards" at some point. :wink:

**Comment by Ragebone** [comment]
off track is actually more correct. I "should" learn for the upcoming exams ....
I will build an "alpha4" but i can't jet say when exactly.

The process looks fairly straight forwoard, reasons for not automating that to get the regular updates implemented?
I'd still let github handle the distribution, so the only cost and effort would be the machine building and then uploading it.
The only problem for me is TeamCity Agents on FreeBSD are a "?".
TeamCity being the main software that i used and misused for many things.
How easy is it to implement a new update branch in FreeNas and is it reasonable?

**Comment by justinclift** [comment]
Well, my reason for not automating the build is probably best described as "laziness". 😀 

It shouldn't be ***too*** hard to implement a new (automatically updated) branch for FreeNAS.  It'd probably end up taking a few hours of time total (spread over a few days), to get it working well.

With TeamCity on FreeBSD, there seems to be a reasonable amount on info on StackOverflow, so it looks like it should work.  Not idea how stable / long term / etc it would work out though. :wink:

**Comment by Ragebone** [comment]
Well, i just forked your repo and added this one line.
https://github.com/Ragebone/freenas-infiniband/commit/807e0fbac62baed2ffef62fc1ef5a875c6d7612b
Legit?  Is that everything i have to do to enable qib?
I have looked through the commits in your repo and went with what i saw there.
Considering that i don't have a clue what i'm doing and how that driver could be named internaly, except from its file names, if it works this evening, i'll make a pullrequest. 😄

**Comment by justinclift** [comment]
Heh Heh Heh.  You're making good progress.

The `device qib` line you added seems like the right thing.  And it sounds like you're getting the hang of building too.

As a thought, I generally built the ISO's using a VM on my OSX desktop (a old Mac Mini).  The VM generally had ~10GB ram, along with plenty of swap space.  Building took um... 7-8 hours I think (from rough memory).  You'll be fine with 16GB ram, just make sure the box has a tonne of free disk space. (like 60GB+ if using ZFS, 100GB+ if you're using UFS)

**Comment by justinclift** [comment]
Just noticed your new commit is in the "fn9_stable" profile.  I think that means you'll need to use `PROFILE=fn9_stable` when doing the build, not `PROFILE=freenas9`.

**Comment by Ragebone** [comment]
I don't know where to change that "PROFILE" setting.  
Github search through the repo didn't turn up anything. 
Is it maybe a make setting or something like that? 
Haven't read through those instructions jet.
EDIT: i just added it to the other profile too. Meh, waht does it matter.

I'm still in the progress of setting everything up.  64GB of RAM are on their way 😄 
Kable is its way back for RMA.
FreeBSD is installed, next is configuring SSH and such.
But i have to hold myself back, my main scedule is learning and finishing exams till the 8th of March.
I may get something done in the spare time but, don't count on it.
Though i will report any progress.

**Comment by justinclift** [comment]
Yep, completely understand.  Good prioritisation is important. This IB stuff is an "in my spare time" thing for me too.

Good luck with your exams, hope they go well. :smile:

**Comment by justinclift** [comment]
Btw:

> I don't know where to change that "PROFILE" setting.

$ make checkout PROFILE=freenas9
$ make release PROFILE=freenas9

The "PROFILE=foo" line tells the build system which of the profile subdirectories to grab the files from.

**Comment by Ragebone** [comment]
caused a mess in my fork trying to get everything setup and together.
It's compiling now. Will let it run over night.
Cloning and setting up the build-environment took half an hour.
Bets on how long this will take?    Xeon e5 2670 and 64GB ram. :smile: 
I guess 2 hours.
EIDIT Wrong!  10 minutes.
Build failed.

Ah i know why.

**Comment by justinclift** [comment]
Hmmm, I'll put in a guess of um... ~3 hours. :smile:

**Comment by Ragebone** [comment]
Well,  PROFILE=freenas9 and fn9_stable crashes 20min into it with this message posted above.
how to debug that?

**Comment by justinclift** [comment]
Ugh, that's unfortunate. 😦 

Looking at the history of the `sys/ofed/drivers/infinband/hw` directory, it seems like only the `mlx4` and `mthca` drivers actually got in:

&nbsp; &nbsp; https://github.com/freenas/os/commits/freebsd11/sys/ofed/drivers/infiniband/hw

Which is weird, as in the commit you pointed out, there were definitely others:

&nbsp; &nbsp; https://lists.freebsd.org/pipermail/svn-src-projects/2011-January/003310.html

**Comment by justinclift** [comment]
As a potential backup plan, it's reasonably common to find Mellanox cards on Ebay fairly cheaply.  eg US$30 at times

**Comment by Ragebone** [comment]
Thanks for that, though i'm already on lookout for a while. Since LTTs youtube video about InfiniBand to be honest.

I have made progress, at least a little.
At some point in time between ca. 2010 and 2013, the driver/hw/ directory of the FreeBSD sources contained source code of the qib driver.
The Github repository, does not contain those, even from the start.
 [This is the first commit, importing OFED in 2011](https://github.com/freebsd/freebsd/commit/5115240a6cdc054f7eea804355742f97c74578d8)
Maybe i'm just to stupid to find the files in and from that stage, but browsing the same directory, at the time of that commit, shows that the hw directory didn't contain the qib driver.
 [This SVN entry shows, there were qib  files!](https://lists.freebsd.org/pipermail/svn-src-projects/2011-January/003310.html)
I tried to get them, but its kind of hard, they are apparently not Github, i cant go that back in GitKraken, and more problems, and problems and problems.

[SVN to the rescue.](https://svnweb.freebsd.org/base?view=revision&revision=216913)
Just needs data extraction!
Test compile, hope they didn't just threw files in there and after that, a diff with the new commits and files from the Linux kernel to get all the patches and updates, and then it hopefully compiles into a final package.

**Comment by justinclift** [comment]
> At some point in time between ca. 2010 and 2013, the driver/hw/ directory of the FreeBSD sources contained source code of the qib driver.

Check your email. :wink:

**Comment by Ragebone** [comment]
i have 3 emails bound to Github.  Checked all of em, nothing. ?

**Comment by justinclift** [comment]
That's weird.  I emailed normangoehl AT hotmail DOT de.  Is that correct?

**Comment by Ragebone** [comment]
I found it, it was buried deep in spam and the Topic was  "Ping". Lets just say i overlooked it.
You answered while i was editing the comment above :smile:

**Comment by justinclift** [comment]
:smile:

**Comment by Ragebone** [comment]
Well,
new cable arrived, and it still doesn't work. So back to fiddling with the software.
Tried to install the OFED Package.   Nope.   The install script crashes on every possible occasion. 
I wan't to swear a lot right now! 

Just to give an example:
error: package opensm is not installed

Yes, its failing the uninstall because stuff isn't installed.     What the F!**&%$%/&%%/$%$§%§$%$§"§%$$%

started to look  deeper into porting stuff from Linux to FreeBSD and i read that someone from  Mellanox did some code shimming, to nearly use the Linux code.

Threw all those files into a git repo to get an overview of what was changed, but i didn't get far with that jet.

**Comment by Ragebone** [comment]
Hey, 
got the Cards working! Finally !
Now i can focus on the driver, knowing that its not some faulty hardware bullshit.

How's your progress ?  FreeNAS Coral infiniband support incoming ?

**Comment by justinclift** [comment]
> got the Cards working! Finally !

Excellent, that definitely is good news. :smile:

> How's your progress ? FreeNAS Coral infiniband support incoming ?

Not looking at it, nor even really thinking about it much.  I'm more focused on getting on [the code](https://github.com/sqlitebrowser/dbhub.io) for [dbhub.io](https://dev1.dbhub.io) working well enough that it's useful for people.

Needing to start generating some revenue in the next couple of months, so I've decided putting time into the IB stuff would just distract me from that in an unhelpful (to me) way. :wink:
